### PR TITLE
Lip 003 - Easy Track

### PR DESCRIPTION
**Pull request by grstepanov**

### CODE REVIEW

File: LIPS/lip-3.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
+| lip: | title: | status: | author: | discussions-to: | created: | updated: |
+| --- | --- | --- | --- | --- | --- | --- |
+| 3 | Easy Track Motions for DAO routine operations | WIP | Bogdan Kovtun (psirex), Gregory Stepanov (grstepanov) | https://research.lido.fi/t/lip-3-easy-track-motions-v2/794 | 2021-07-09 | |
+
+# Summary
+
+## Problem
+
+Lido DAO governance currently relies on Aragon Voting App model. This means DAO approves or rejects proposals via direct governance token voting. Though transparent and reliable, it is not a convenient way to make decisions only affecting small groups of Lido DAO members. Besides, direct token voting doesn't exactly reflect all the decision-making processes within the Lido DAO and is often used only to formalize an existing consensus.
```
Comment (kadmil) [inline LIPS/lip-3.md]: It doesn't seem to be correct that the Lido DAO uses Aragon votings to formalize any decisions — for ones not requiring the on-chain tx we do use snapshots.

File: LIPS/lip-3.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+There are a few natural sub-governance groups within the DAO, e.g. node operators committee, financial operations team and LEGO committee. Every day they need to take routine actions only related to their field of expertise. The decisions they make hardly ever spark any debate in the community, and votings on such decisions often struggle to attract wider DAO attention and thus, to pass.
```
Comment (kadmil) [inline LIPS/lip-3.md]: to get the quorum & be enacted as the on-chain tx

File: LIPS/lip-3.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+## Proposed solution
+
+We propose Easy Track frictionless motions as a solution to this problem. 
+Easy Track motion is a lightweight voting considered to have passed if the minimum objections threshold hasn’t been exceeded. As opposed to traditional Aragon votings, Easy Track motions are cheaper (no need to vote ‘pro’, token holders only have to vote ‘contra’ if they have objections) and easier to manage (no need to ask broad DAO community vote on proposals sparking no debate).
```
Comment (kadmil) [inline LIPS/lip-3.md]: `threshold hasn’t been exceeded` — let's mention that there's a timeframe there?

File: LIPS/lip-3.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+- Funds being allocated into reward programs
+- Funds being allocated to LEGO program
+
+More types of motions can be added later, e.g.:
+- Topping up Lido staking referral program
+- Purchasing staking cover
+
+## Sub-governance groups
+
+There's a sub-governance group within the Lido DAO for each of the voting cases listed above. Those groups have been formed intentionally or in a more natural way, and at this point creating and maintaining periodical Aragon votings falls within their remits.
```
Comment (kadmil) [inline LIPS/lip-3.md]: As mentioned below, we don't have finance ops team / subgroup currently. There's a finance multisig, but it's a hack job and we'll need a proper one set up.

File: LIPS/lip-3.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+The proposed improvement will require formalizing sub-governance groups:
+
+1. **Node Operators Committee** to start staking limit increase motions.
+2. **Financial Operations Team** to start funds allocation motions (i.e. LEGO and reward programs).
+
+Node Operators Committee should include addresses of active node operators partnering with Lido, the list is easy to look up in the [Node Operators Registry](https://github.com/lidofinance/lido-dao/blob/master/contracts/0.4.24/nos/NodeOperatorsRegistry.sol) smart contract. Financial Operations Team should consist of established DAO members' addresses and be represented as a Gnosis Safe Multisig with access to corresponding Easy Track features.
+
+# Specification
+
+To create a feature as flexible and extendable as possible, we propose implementing Easy Track functionality within several contracts rather than building a single contract for it.
```
Comment (kadmil) [inline LIPS/lip-3.md]: The maximal flexibility in itself wasn't a prio, but having enough of it to accommodate for current and future "operational" transactions being regulary made from Lido DAO side. Having opt-out gov thing flexible and extendable in itself sounds properly scary.

File: LIPS/lip-3.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+# Specification
+
+To create a feature as flexible and extendable as possible, we propose implementing Easy Track functionality within several contracts rather than building a single contract for it. 
+> The statement below is abstract, please visit [specification.md](https://github.com/lidofinance/easy-track/blob/master/specification.md) for full detailed specification.
+
+## Easy Track workflow and motion life cycle
+
+A group of core smart contracts will implement basic logic and store data related to Easy Track motions, and a series of separate EVMScriptFactory smart contracts will represent all the supported types of Easy Track motions. 
+This architecture aims to support further extension by adding new types of Easy Track motions without replacing the core of the proposed system.
+1. A motion can be started by calling `createMotion()` function on `EasyTrack` core contract.
```
Comment (kadmil) [inline LIPS/lip-3.md:54]: I'd suggest adding small intro paragraph "that's what regular motion operation looks like" — we're telling the story of overall architecture and then jump to smaller detail right ahead.

File: LIPS/lip-3.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+## Easy Track workflow and motion life cycle
+
+A group of core smart contracts will implement basic logic and store data related to Easy Track motions, and a series of separate EVMScriptFactory smart contracts will represent all the supported types of Easy Track motions. 
+This architecture aims to support further extension by adding new types of Easy Track motions without replacing the core of the proposed system.
+1. A motion can be started by calling `createMotion()` function on `EasyTrack` core contract. 
+2. Upon motion creation, `EasyTrack` calls `createEvmScript()` function on the corresponding `EVMScriptFactory` contract passing all the data required to generate a motion enactment EVM script.
+3. `EVMScriptFactory` smart contract generates the script and returns it to the `EasyTrack` core contract.
+4. `EasyTrack` contract conducts the motion according to the motion settings (i.e. motion duration, objections threshold and maximum active motions count). While active, the motion can be canceled at any time by the address that has started it.
```
Comment (kadmil) [inline LIPS/lip-3.md]: May want to mention explicitly that the motion could be cancelled by the address started it at any time before the execution

File: LIPS/lip-3.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+5. As soon as the motion duration expires, it becomes possible to enact the motion. Anyone can enact the motion from that moment on.
+6. To enact the motion, the `EasyTrack` contract recreates the enactment EVM script and passes it to the `EVMScriptExecutor` smart contract for execution.
+7. `EVMScriptExecutor` holds all the permissions to run valid enactment EVM scripts passed from the `EasyTrack` smart contract (or Aragon voting app).
```
Comment (kadmil) [inline LIPS/lip-3.md]: Do we need Aragon voting mentioned here? Even if so, a separate point could be a bit better, imo

File: LIPS/lip-3.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+### `EVMScriptFactoriesRegistry` contract
+
+`EVMScriptFactoriesRegistry` contract provides methods to add/remove `EVMScriptFactory` into the list of allowed `EVMScriptFactory` contracts and also an internal method to create EVM scripts. `EVMScriptFactoriesRegistry` inherits from `EasyTrackStorage`.
+The internal method `_createEVMScript()` creates the enactment EVM script by calling the corresponding `EVMScriptFactory` contract and also checks if the resulting script meets all the required script factory permissions.
+
+### `EasyTrackStorage` contract
+
+`EasyTrackStorage` contract stores all the variables of Easy Track system. It inherits from battle-tested `Initializable`, `PausableUpgradeable` and `AccessControlUpgradeable` OpenZeppelin contracts.
+On the other hand, `MotionSettings` and `EVMScriptFactories` inherit from `EasyTrackStorage` to make variables stored in the latter accessible to their methods.
```
Comment (kadmil) [inline LIPS/lip-3.md]: If those are two different contracts, would we be forced to have two sets of variables, which should be just the same? This sounds a bit unclear & may cause ops hazard.

File: LIPS/lip-3.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+______
+## `EVMScriptFactory` smart contracts
+
+Initially, five `EVMScriptFactory` contracts will represent five Easy Track motion types.
```
Comment (kadmil) [inline LIPS/lip-3.md:95]: We've mentioned three types on the start; I know that we'll have all five, but here it's worth explaining the difference, imo

File: LIPS/lip-3.md [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+- `RewardProgramsRegistry` already contains the reward program address.
+
+#### `TopUpRewarProgram` script factory contract
+
+`TopUpRewarProgram` contract creates EVM Script to top up balances of reward programs. Addresses to be topped up are restricted to the list stored in the `RewardProgramsRegistry` contract.
+
+## `EVMScriptExecutor` contract
+
+The EVM scripts generated by the script factory contracts can be executed by a separate `EVMScriptExecutor` contract. It implements only `executeEVMScript()` method and accepts scripts built following the Aragon's EVM script structure. Current implementation uses [CallsScript.sol](https://github.com/aragon/aragonOS/blob/next/contracts/evmscript/executors/CallsScript.sol) contract of the Aragon framework. 
+`EVMScriptExecutor` permissions list is the same as the one of Aragon Voting App. While sharing the EVM script structure, both `EasyTrack` contract and Aragon Voting App are able to execute scripts using `EVMScriptExecutor` contract.
```
Comment (kadmil) [inline LIPS/lip-3.md]: Would it have the same roles assigned from the get-go? If so, let's make it more explicit?

File: LIPS/lip-3.md [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+___
+# Sanity and security
```
Comment (kadmil) [inline LIPS/lip-3.md]: Maybe `security and sanity checks`?

File: LIPS/lip-3.md [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+The proposed Easy Track Motions design involves timelock as a native security feature. Any suspicious motion can be rejected or canceled within a relatively long motion duration period. To prevent malicious actions, we propose putting the following additional limitations in place (numeric values fixed based on community feedback from v1):
```
Comment (kadmil) [inline LIPS/lip-3.md]: Do we have a link on the feedback?

File: LIPS/lip-3.md [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+- It should be impossible to set objections threshold at more than 5% of the total LDO supply. Nevertheless, the default objections threshold should be set a lot lower at 0.5% of the total LDO supply.
+- It should be impossible to set motion duration at less than 48 hours. For any motion, there should be enough time for DAO to submit objections and reject the motion. 
+- It should be impossible to spam motions. The default limit for simultaneously active motions will be set at 12 and it can only be increased by the DAO to up to 24 motions at a time.
```
Comment (kadmil) [inline LIPS/lip-3.md]: Let's add explicitly that the limit is on all types of motions in total, and not for every time in particular?

**Review by kadmil — COMMENTED** [review]
Added a couple nitpicks, but overall seems good! I'd rephrase the intro part & focus on safety measures & rules we do have around 1) running current types of easy tracks; 2) changing available types of Easy Tracks (btw, how are new entities allowed to creating Easy Tracks?); 3) roles management & connection to Aragon setup we do have.

File: LIPS/lip-3.md (hunk 1 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Not any, but I believe the Lido DAO does **often** formalize existing consensus via Aragon votings. 
That's what happens with the rewards allocations: first the community sentiment gets measured on the research forum and then rubber-stamped with the omnibus. In fact, it's something already decided but still requiring an on-chain confirmation.

File: LIPS/lip-3.md (hunk 4 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: I think the finance multi-sig owners may be considered an informal sub-governance group at this point. Adding this into the LIP to make it more precise.

File: LIPS/lip-3.md (hunk 5 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Like I put it in the LIP, flexibility and extendability were the two primary reasons why we preferred splitting the feature into multiple contracts rather than building a Frankenstein master contract instead. Another reason worth mentioning is having the feature supervised, adding it into the proposal.

File: LIPS/lip-3.md (hunk 10 above)
Comment (grstepanov) [reply LIPS/lip-3.md:95]: I don't see a contradiction here, as we mentioned three types of votings we would like to wrap into motions. Wrapping resulted in five motions, which is pretty obvious from the list right below this line.

File: LIPS/lip-3.md (hunk 13 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Most of this discussion happened on research forum [here](https://research.lido.fi/t/lip-3-easy-track-motions-for-dao-routine-operations/680).
We have a discussion link in the LIP header already, but I am adding it here as well.

File: LIPS/lip-3.md (hunk 3 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Done.

File: LIPS/lip-3.md (hunk 6 above)
Comment (grstepanov) [reply LIPS/lip-3.md:54]: Added.

File: LIPS/lip-3.md (hunk 7 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Done.

File: LIPS/lip-3.md (hunk 8 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Removed the Aragon mention, covering it in a new paragraph about ET<>Aragon relations.

File: LIPS/lip-3.md (hunk 12 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Done.

File: LIPS/lip-3.md (hunk 14 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Good point.

File: LIPS/lip-3.md (hunk 11 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Specified in the Roles management and connection to the current Aragon setup paragraph

File: LIPS/lip-3.md (hunk 9 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Actually, this contract has been removed alongside the upgradeability, dropping this paragraph altogether.
The variables are now stored in MotionSettings, EVMScriptFactoriesRegistry and EasyTrack contracts right next to the logic.

File: LIPS/lip-3.md (hunk 1 above)
Comment (kadmil) [reply LIPS/lip-3.md]: In that case, let's add something along the lines of "reached in discussion on research forum or snapshot vote", please. We don't just slap things into Aragon votes, are we?

File: LIPS/lip-3.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+- Funds being allocated into reward programs
+- Funds being allocated to LEGO program
+
+More types of motions can be added later, e.g.:
+- Topping up Lido staking referral program
+- Purchasing staking cover
+
+## Sub-governance groups
+
+There's a formal or an informal sub-governance group within the Lido DAO for each of the voting cases listed above. Those groups have been formed intentionally or in a more natural way, and at this point creating and maintaining periodical Aragon votings falls within their remits.
```
Comment (kadmil) [inline LIPS/lip-3.md]: For the last part, it's not creating the votings, but ensuring the necessary transactions get processed through means of Aragon vote. The groups don't start the votes in general, and only manage actions required for their respective ops

File: LIPS/lip-3.md (hunk 7 above)
Comment (kadmil) [reply LIPS/lip-3.md]: still not clear whether pending (timeframe passed, not enough reject-votes to reject, not executed) motion could be cancelled by the creator

File: LIPS/lip-3.md [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+- Motion enactment. If the objections threshold hasn't been exceeded within or after the voting period, the motion can be enacted by anyone. Before enactment, the motion can be canceled by the address that started it.
+
+This is how it works in terms of smart contract interactions:
+1. A motion can be started by calling `createMotion()` function on `EasyTrack` core contract. 
+2. Upon motion creation, `EasyTrack` calls `createEvmScript()` function on the corresponding `EVMScriptFactory` contract passing all the data required to generate a motion enactment EVM script.
+3. `EVMScriptFactory` smart contract generates the script and returns it to the `EasyTrack` core contract.
+4. `EasyTrack` contract conducts the motion according to the motion settings (i.e. motion duration, objections threshold and maximum active motions count). While active, the motion can be canceled at any time by the address that has started it.
+5. As soon as the motion duration expires, it becomes possible to enact the motion. Anyone can enact the motion from that moment on.
```
Comment (kadmil) [inline LIPS/lip-3.md]: Need a point on what happens if objection threshold is reached

File: LIPS/lip-3.md [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+The proposed Easy Track Motions design involves timelock as a native security feature. Any suspicious motion can be rejected or canceled within a relatively long motion duration period. To prevent malicious actions, we propose putting the following additional limitations in place (numeric values fixed based on [community feedback from v1](https://research.lido.fi/t/lip-3-easy-track-motions-for-dao-routine-operations/680)):
+
+- It should be impossible to set objections threshold at more than 5% of the total LDO supply. Nevertheless, the default objections threshold should be set a lot lower at 0.5% of the total LDO supply.
+- It should be impossible to set motion duration at less than 48 hours. For any motion, there should be enough time for DAO to submit objections and reject the motion. 
+- It should be impossible to spam motions. The default limit for simultaneously active motions will be set at 12 and it can only be increased by the DAO to up to 24 motions at a time. This value refers to total number of active motions of all types.
+- Pause lever should be added. It is proposed to set up an Easy Track security multi-sig to pause Easy Track in case of emergency.
```
Comment (kadmil) [inline LIPS/lip-3.md]: May want to specify what the pause entails: no new motions, active & pending can't be executed, ones passing during the pause can't be executed after unpause (or what the exact rules are?)

File: LIPS/lip-3.md (hunk 15 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Fixed.

File: LIPS/lip-3.md (hunk 16 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Added the point.

File: LIPS/lip-3.md (hunk 17 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Specified what pause entails.

File: LIPS/lip-3.md (hunk 1 above)
Comment (grstepanov) [reply LIPS/lip-3.md]: Fixed wording.

**Review by kadmil — APPROVED** [review]
LGTM
